manpages/imodfillin.man
.na
.nh
.TH imodfillin 1 2.6.2 BL3DEMC
.SH NAME
imodfillin \- fill in missing contours in IMOD model using mesh data
.SH SYNOPSIS
imodfillin  [options]  input_model  output_model
.SH DESCRIPTION
imodfillin will fill in contours in an IMOD model on sections that were
skipped during modeling.  It interpolates a contour between existing ones
using information in the object mesh
created by imodmesh(1).  Each interpolated contour will fall on the surface
connecting existing contours.  It is thus necessary to run imodmesh(1) before
imodfillin.  Since you need to have a mesh between contours on non-adjacent 
sections, you should use one or more imodmesh options to connect across gaps in
the data.  If sections have either a complete set of contours or no contours 
at all, then the
.B -s
option will suffice.  If the pattern is less rigid than this, use the 
.B -P
option in addition (or exclusively), where the
number of passes that you specify with this option 
equals the maximum number of sections of separation between contours that
you wish to connect.  Do not use any imodmesh option that would set the z 
increment
to greater than 1 (i.e.,
.B -l, -i,
or
.B -z
), for then imodfillin will probably create contours on sections that
already have them.
.P
By default, the program will attempt to fill in all closed-contour objects
that have meshes, but will skip open-contour objects.  Use the 
.B -o
option to specify a subset of objects to be done, or to force it to do 
particular open-contour objects.
.P
The created contours will inherit any surface or time information contained
in the meshes.  If your original contours have surface numbers, be sure to
mesh with the 
.B -S
option to preserve this surface information in the meshes, so that the
interpolated contours will have the same surface numbers as the original ones.
.SH OPTIONS
.TP
.B -e
Place new contours in existing objects.  Use this option if you do not
plan to check and modify the interpolated contours.
.TP 
.B -n
Place new contours in new objects; this is the default, so that contours
can be checked conveniently.  One new object
will be created for each object that is being filled in.  These contours
may be moved into the proper object either one by one or all at once via
the Edit-Contour-Move dialog and M hotkey in 3dmod(1).
.TP
.B -o \fIlist\fR
Do operations only on objects in the list of numbers.  The list can consist
of comma-separated ranges (e.g., 1,7-11,13-17,19).
Without this option, all closed contour objects with meshes are filled in.
.TP
.B -i \fIvalue\fR
Set the Z increment for filling in contours.  This option has two effects:
the program will fill in only gaps that are wider than the given value,
and it will insert new contours at this given Z increment instead of on every
section.
.TP
.B -R \fIvalue\fR  
Resolution or tolerance for point reduction
in the generated contours.  Points are eliminated if they are within this
distance of the lines connecting the remaining points, just as with the -R
option in imodmesh(1).
Values in the range 0.25 to 0.5 are often useful.
.SH FILES
If the model file already exists, it becomes a backup file with ~ added to its
name.
.SH AUTHOR
David Mastronarde
.SH SEE ALSO
3dmod(1), imodmesh(1)
.SH BUGS
Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names
will differ in different Java versions.  This bug will not be fixed because it
does not appear to cause any harm.  Only the automated tests where modified.
(Bug# 1753)


imodutil/imodfillin.c
/*
 *  imodfillin.c --  Fills in missing contours based on mesh data
 *
 *  Author: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 *  Log at end of file
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "imodel.h"
#include "b3dutil.h"

void fillin_from_mesh(Imod *imod, int ob, int newobj, int zinc, float tol);
#define DEFAULT_TOL 0.25

static void imodfillin_usage(char *name, int retcode)
{
  printf("usage: %s [options] <input model> <output model>\n", name);
  printf("options:\n");

  printf("\t-e\tPlace new contours in existing objects.\n");
  printf("\t-n\tPlace new contours in new objects (default).\n");
  printf("\t-o list\tFill in the given list of objects only (ranges allowed).\n");
  printf("\t-i #\tFill in only gaps bigger than the given Z increment.\n");
  printf("\t-R #\tTolerance (maximum error) for point reduction"
         " (default %.2f).\n", DEFAULT_TOL);
  exit(retcode);
}

int main( int argc, char *argv[])
{
  int  i, c, ob;
  int  zinc = 1;
  int  newobj = TRUE;
  float tol = DEFAULT_TOL;
  int  obj_list_size = 0;
  int  *obj_list;
  Imod *model;
  int  obsave, cosave, ptsave;
  int doit;
  int origsize;
  int anymesh = 0;
  char *progname = imodProgName(argv[0]);

  if (argc == 1){
    imodVersion(progname);
    imodCopyright();
    imodfillin_usage(progname, 0);
  }

  for (i = 1; i < argc ; i++){
    if (argv[i][0] == '-'){
      switch (argv[i][1]){
               
      case 'i':
        zinc = atoi(argv[++i]);
        if (zinc < 1)
          zinc = 1;
        break;
               
      case 'R':
        tol = atof(argv[++i]);
        if (tol <= 0.001)
          tol = 0.001;
        break;
               
      case 'o':
        obj_list = parselist(argv[++i], &obj_list_size);
        if (!obj_list) {
          fprintf(stderr, "%s: Error parsing object list\n", progname);
          exit(3);
        }
        break;
               
      case 'e':
        newobj = 0;
        break;

      case 'n':
        newobj = 1;
        break;

      case '?':
        imodfillin_usage(progname, -1);
        break;

      default:
        fprintf(stderr, "%s: unknown option %s\n", progname, argv[i]);
        imodfillin_usage(progname, -1);
        break;
               
      }
    } else
      break;
  }

  if (i >= argc - 1)
    imodfillin_usage(progname, -1);

  model = imodRead(argv[argc - 2]);
  if (!model) {
    fprintf(stderr, "%s: Fatal error reading model %s\n", progname, 
            argv[argc - 2]);
    exit(1);
  }

  imodGetIndex(model, &obsave, &cosave, &ptsave);

  /* Loop on objects, finding ones with meshes that are either on the list
     or consist of closed contours */
  origsize = model->objsize;
  for (ob = 0; ob < origsize; ob++) {
    doit = model->obj[ob].meshsize;
    if (doit)
      anymesh = 1;
    if (doit && obj_list_size) {
      doit = 0;
      for (i = 0; i < obj_list_size; i++)
        if (ob + 1 == obj_list[i]) doit = 1;
    } else if (doit) 
      doit = iobjClose(model->obj[ob].flags);

    if (doit) {
      printf("Examining object %d\n", ob + 1);
      fillin_from_mesh(model, ob, newobj, zinc, tol);
    }
  }    

  if (!anymesh)
    printf ("No objects with meshes found; be sure to run imodmesh "
            "with the -s flag\n");

  imodSetIndex(model, obsave, cosave, ptsave);
  imodObjviewComplete(model);

  imodBackupFile(argv[argc - 1]);
  if (imodOpenFile(argv[argc - 1], "wb", model)) {
    fprintf(stderr, "%s: Fatal error opening new model %s\n", progname,
            argv[argc - 1]);
    exit (1);
  }
  imodWriteFile(model);
  exit(0);
}

void fillin_from_mesh(Imod *imod, int ob, int newobj, int zinc, float tol)
{
  Iobj *obj = &imod->obj[ob];
  Iobj *destobj = obj;   /* Destination object */
  int resol;         /* resolution for highest res mesh */
  int me, i, ind;      /* indices */
  int *listp;        /* pointer to normal-vertex indices */
  int ninpoly;       /* # of vertexes in polygon */
  Ipoint *vertp;     /* pointer to normal-vertex list */
  float zmax, zmin;  /* max and min z in polygon */
  int ntriang;       /* # of triangles in polygon */
  int zadd;          /* Z value at which to add contour */
  int firstv;        /* index of first vertex in polygon */
  Icont *cont;       /* contour being added to */
  int coadd;         /* number of that contour */
  float red, green, blue;
  int listInc, vertBase, normAdd;

  imodMeshNearestRes(obj->mesh, obj->meshsize, 0, &resol);
  for (me = 0; me < obj->meshsize; me++) {
    if (imeshResol(obj->mesh[me].flag) != resol) 
      continue;
     
    listp = obj->mesh[me].list;
    vertp = obj->mesh[me].vert;
    i = 0;
    while (i < obj->mesh[me].lsize) {
      if (imodMeshPolyNormFactors(listp[i++], &listInc, &vertBase, &normAdd)) {
                    
        /* Find min and max Z in polygon, first vertex index and
           number of vertices */
        firstv = i + vertBase;   /* first vertex index */
        zmin = vertp[listp[firstv]].z;
        zmax = zmin;
        ninpoly = 0;
        while (listp[i] != IMOD_MESH_ENDPOLY) {
          ind = listp[i + vertBase];
          i += listInc;
          if (vertp[ind].z < zmin)
            zmin = vertp[ind].z;
          if (vertp[ind].z > zmax)
            zmax = vertp[ind].z;
          ninpoly++;
        }
                    
                    
        /* If difference in Z is not > than desired increment, 
           skip */
        if (zmax - zmin <= zinc) 
          continue;
                    
        /* See if need to make a new object now */
        if (newobj) {
          imodNewObject(imod);
          obj = &imod->obj[ob];
          destobj = &imod->obj[imod->objsize - 1];
          /* Copy properties, then restore new color and clear 
             out data */
          red = destobj->red;
          green = destobj->green;
          blue = destobj->blue;
          imodObjectCopy(obj, destobj);
          destobj->red = red;
          destobj->green = green;
          destobj->blue = blue;
          destobj->cont = NULL;
          destobj->mesh = NULL;
          destobj->contsize = 0;
          destobj->meshsize = 0;
          destobj->store = NULL;
          destobj->meshParam = NULL;
          printf("Adding new contours to new object %d\n", 
                 imod->objsize);
          newobj = 0;
        }
                    
        zadd = floor((double)(zmin + zinc + 0.001));
        ntriang = ninpoly / 3;
        while (zadd < zmax) {
          cont = imodContourNew();
          coadd = imodObjectAddContour(destobj, cont);
          if (coadd < 0) {
            fprintf(stderr, "Fatal error: cannot get new "
                    "contour or add it to object");
            exit(3);
          }
          free(cont);
          cont = &destobj->cont[coadd];
          cont->surf = obj->mesh[me].surf;
          cont->time = obj->mesh[me].time;
                         
          imodMeshInterpCont(listp, vertp, ntriang, firstv, listInc, zadd, cont);
          imodContourReduce(cont, tol);
          zadd += zinc;
        }
      }
    }
  }
}

     
/*

  $Log$
  Revision 3.13  2008/04/04 21:21:28  mast
  Free contour after adding to object

  Revision 3.12  2006/09/12 15:02:42  mast
  rename mesh members, null out meshparam

  Revision 3.11  2006/06/26 14:48:49  mast
  Added b3dutil include for parselist

  Revision 3.10  2005/09/11 19:22:11  mast
  Changes for new style of mesh

  Revision 3.9  2005/02/11 01:42:33  mast
  Warning cleanup: implicit declarations, main return type, parentheses, etc.

  Revision 3.8  2004/11/05 19:05:29  mast
  Include local files with quotes, not brackets

  Revision 3.7  2004/07/07 19:25:30  mast
  Changed exit(-1) to exit(3) for Cygwin

  Revision 3.6  2003/10/26 14:46:41  mast
  fixed problem in eliminating getopt

  Revision 3.5  2003/10/24 03:05:23  mast
  open as binary, strip program name and/or use routine for backup file

  Revision 3.4  2003/08/01 00:16:14  mast
  Complete object views when saving model, since there may be new objects

  Revision 3.3  2003/02/21 23:15:33  mast
  Open new file as wb

  Revision 3.2  2002/06/21 00:26:03  mast
  Needed to swap points from first triangle of a strip if they were out of
  order relative to the following points of the strip

  Revision 3.1  2001/12/17 18:53:55  mast
  First version of program

*/
